### ISSUE TITLE
Request connext collateral based on order amount

### ISSUE THREAD

**Issue by kilrau** [ExchangeUnion/xud#1845]
As discussed before, we should request channel collateralization (remote balance) from the connext node ahead of (swap) time to avoid on-the-fly collateralization.

How it could work:
1. `placeOrder`
2. `xud` checks `tradinglimits`
3. if lnd remote balance not enough to fully swap order, xud fail order placement as is
4. if connext remote balance is not enough to fully swap, xud request remainder as collateral from the node and puts order on hold
5. once collateralization is done (xud can just hit the `/balance` endpoint to see if remote balance is there
6. once remote balance appears, place order. If remote balance doesn't appear within certain timeout, drop order. (some tbd's here, e.g. how to handle arby replacing current order on-hold)
7. subsequent replace orders (e.g. via arby replace) should pass without collateralization needed

Depending on:
https://github.com/connext/rest-api-client/issues/89 -> PR: https://github.com/connext/rest-api-client/pull/90
https://github.com/connext/indra/issues/1405#issuecomment-681928062 -> PR: https://github.com/connext/indra/pull/1406

**Comment by kilrau** [comment]
https://github.com/connext/indra/issues/1405#issuecomment-681928062 + https://github.com/connext/rest-api-client/issues/89 should get us all we need. Connext moved fast, and a PR for https://github.com/connext/indra/issues/1405 is already up: https://github.com/connext/indra/pull/1406 props to @rhlsthrm

**Comment by erkarl** [comment]
> As discussed before, we should request channel collateralization (remote balance) from the connext node ahead of (swap) time to avoid on-the-fly collateralization.
> 
> How it could work:
> 
>     1. `placeOrder`
> 
>     2. `xud` checks `tradinglimits`
> 
>     3. if lnd remote balance not enough to fully swap order, xud fail order placement as is
> 
>     4. if connext remote balance is not enough to fully swap, xud request remainder as collateral from the node and puts order on hold

What if instead of putting it on hold we fail the `PlaceOrder` request and ask the user to try again later. Consecutive, `PlaceOrder` requests would fail immediately if the request for collateral is in progress (and additional collateral is needed to fill that order).

>     7. subsequent replace orders (e.g. via arby replace) should pass without collateralization needed

Hmm, replace order could also need additional collateral if the price changes.

**Comment by kilrau** [comment]
> What if instead of putting it on hold we fail the `PlaceOrder` request and ask the user to try again later (while starting the collateral request to the node). Consecutive, `PlaceOrder` requests would fail immediately if the request for collateral is in progress (and additional collateral is needed to fill that order).

That would also work. Easier from an implementation perspective I guess (even though there would be a new check against "collateralization already in progress, don't re-request").

> Hmm, replace order could also need additional collateral if the price changes.

If the price changes and consequently amount exceeds remote balance. What we can do is, request 3% more from the node to account for that.

**Comment by erkarl** [comment]
> If the price changes and consequently amount exceeds remote balance. What we can do is, request 3% more from the node to account for that.

Yep, some buffer would probably be nice to have.

 The scenario I was thinking of was when the user deposits additional funds to BTC/LTC channels/CEX. In that case the logic would be the same as regular `PlaceOrder` - we wait until the additional collateralization is added in the background.

**Comment by erkarl** [comment]
@sangaman any thoughts/ideas?

**Comment by kilrau** [comment]
Reiterating @raladev 's thoughts on this on yesterday's call: I think we should solely rely on order amounts to request connext collateral. Connext gives us this flexibility, so we should use it. That means in particular, xud would **not** request and collateral based on channel opening in boltz-style, e.g. I deposit 100 USDT into a channel, node automatically collateralizes 25 USDT, no matter if orders are up or not.

**Comment by sangaman** [comment]
> The scenario I was thinking of was when the user deposits additional funds to BTC/LTC channels/CEX. In that case the logic would be the same as regular `PlaceOrder` - we wait until the additional collateralization is added in the background.

I agree with this logic. Failing rather than waiting if the collateral/remote balance isn't sufficient, particularly since delaying an order isn't great UX given that prices may be changing and the order may no longer be desirable after collateral is there.

> That means in particular, I suggest xud not to request and collateral based on openchannel, but on placeorder only.

I think I'd be OK with this, but then would we expect that the first order that receives via Connext will always fail due to needing to wait for collateralization? Maybe it can be optional when using `OpenChannel` with connext?

**Comment by kilrau** [comment]
> but then would we expect that the first order that receives via Connext will always fail due to needing to wait for collateralization?

No, that's precisely what we would want to prevent here, at least failing as in "swap timeout failing". In the moment the first order receiving connext is placed, xud requests collateralization for the required amount. Question is if we fail the order on *CLI level*  ([@erkarl  s suggestion](https://github.com/ExchangeUnion/xud/issues/1845#issuecomment-682589027)) in that case or put it on hold in the background and automatically add it to the order book once the collateralization is through. The latter is much better from user perpective, but would be more complicated to implement I guess. For example we'd need a timer to fail the order, if collateralization is taking longer than X minutes, because then it might be undesirable for the user that the order goes through.

**Comment by sangaman** [comment]
> No, that's precisely what we would want to prevent here, at least failing as in "swap timeout failing". In the moment the first order receiving connext is placed, xud requests collateralization for the required amount. Question is if we fail the order on CLI level (@erkarl s suggestion) in that case or put it on hold in the background and automatically add it to the order book once the collateralization is through. The latter is much better from user perpective, but would be more complicated to implement I guess. For example we'd need a timer to fail the order, if collateralization is taking longer than X minutes, because then it might be undesirable for the user that the order goes through.

Yes, I was referring to the order being rejected upon being submitted due to lack of collateral, not due to failing a swap. I'm not sure whether adding the order on a delay is better tho as I mentioned above since a user might only want the order to be good for a few seconds, particularly in the case of market orders.

**Comment by erkarl** [comment]
> I'm not sure whether adding the order on a delay is better tho as I mentioned above since a user might only want the order to be good for a few seconds, particularly in the case of market orders.

I'm strongly of the opinion that we shouldn't automatically place the order in the orderbook once the collateralization completes.

My desired flow would be:
```
sell 0.1 BTC/USDT 11000
-> CLI: started collateralization of your channel, please try again in 1 minute
30 sec later
sell 0.1 BTC/USDT 11000
-> CLI: collateralization already in progress, please try again in 1 minute
once the collateralization completes
-> xud logs print collateralization complete, new inbound balance: 12345
sell 0.1 BTC/USDT 11000
order placed to the orderbook
```

**Comment by erkarl** [comment]
Or alternatively the behaviour that @kilrau suggests would be behind a flag:
`sell 0.1 BTC/USDT 11000 --wait-for-collateralization-and-then-place-order-automatically`

**Comment by kilrau** [comment]
> a user might only want the order to be good for a few seconds, particularly in the case of market orders.

> I'm strongly of the opinion that we shouldn't automatically place the order in the orderbook 

Ok I see that. Let's go with Karl's [suggestion](https://github.com/ExchangeUnion/xud/issues/1845#issuecomment-684981767). Regarding wording/structure of output I suggest:
```
sell 0.1 BTC/USDT 11000
-> CLI: requested inbound liquidity for your USDT channel, please try again in ~1 minute
30 sec later
sell 0.1 BTC/USDT 11000
-> CLI: already requested inbound liquidity for your USDT channel, please try again in ~1 minute
once the collateralization completes
sell 0.1 BTC/USDT 11000
-> matched 0.1 BTC @ 11000 with peer FunnyGuy order 4522e390-ec73-11ea-b021-9d791f14027a, attempting swap...
(regular matching output)
```

**Comment by raladev** [comment]
> > a user might only want the order to be good for a few seconds, particularly in the case of market orders.
> 
> > I'm strongly of the opinion that we shouldn't automatically place the order in the orderbook

may we add some flag to autoplace order after collateralization? (replacing is annoying)

**Comment by kilrau** [comment]
> may we add some flag to autoplace order after collateralization? (replacing is annoying) @kilrau @erkarl

We discussed before that this can be dangerous because the market may move significantly within the time the collateralization takes. As an optional flag this can be useful though (for people who understand the consequences). Other opinions?

**Comment by raladev** [comment]
> > may we add some flag to autoplace order after collateralization? (replacing is annoying) @kilrau @erkarl
> 
> We discussed before that this can be dangerous because the market may move significantly within the time the collateralization takes. As an optional flag this can be useful though (for people who understand the consequences). Other opinions?

Yeap, i am talking about optional flag with false value as default (no autoplacement as default).

**Comment by kilrau** [comment]
I think I'd be ok with it, let's see what @erkarl and @sangaman think

**Comment by erkarl** [comment]
> I think I'd be ok with it, let's see what @erkarl and @sangaman think

I'm also OK with an optional flag that enables this functionality. This can be done outside of this PR as a follow-up.
